### PR TITLE
[Fixes #7518] NullReferenceException thrown when Controller method us…

### LINKED ISSUE #7518: NullReferenceException thrown when Controller method uses Guid parameter default value
I'm using ASP.NET Core 2.0 with MVC and testing my app within Visual Studio 2017(15.3.0) using IIS Express.  I'm also using version 7.1 of the C# language.

When I browse to a route that is associated with the following type of controller method using attribute routing:
```csharp
[HttpGet]
public Task<IActionResult> List([FromQuery]Guid hookId = default)
{
     return Task.FromResult<IActionResult>(Ok());
}
```

I get the following exception
```
NullReferenceException: Object reference not set to an instance of an object.
  lambda_method(Closure , object , Object[] )
  Microsoft.Extensions.Internal.ObjectMethodExecutor.Execute(object target, Object[] parameters)
  Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker+<InvokeActionMethodAsync>d__12.MoveNext()
  System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
  System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker+<InvokeNextActionFilterAsync>d__10.MoveNext()
  System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
  Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Rethrow(ActionExecutedContext context)
  Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
  Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker+<InvokeInnerFilterAsync>d__14.MoveNext()
  System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
  System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker+<InvokeNextResourceFilter>d__22.MoveNext()
  System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
  Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)
  Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
  Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker+<InvokeFilterPipelineAsync>d__17.MoveNext()
  System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
  System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker+<InvokeAsync>d__15.MoveNext()
  System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
  System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  Microsoft.AspNetCore.Builder.RouterMiddleware+<Invoke>d__4.MoveNext()
  System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
  System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
```

If I change the parameter of type Guid to be of type Guid? instead everything works as expected:
```csharp
[HttpGet]
public Task<IActionResult> List([FromQuery]Guid? hookId = default)
{
     return Task.FromResult<IActionResult>(Ok());
}
```

I was expecting the routing to work with a default Guid parameter value or if it wasn't supported to throw a more meaningful exception message.

### PR DESCRIPTION
**Pull request by kichalla**
…es Guid parameter default value

Note that this is for an RC1 issue and so review it per your priorities.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ParameterDefaultValues.cs [hunk 1]
```diff
@@ -21,34 +21,38 @@ public static object[] GetParameterDefaultValues(MethodInfo methodInfo)
...
+        }
 
-                if (parameterInfo.HasDefaultValue)
+        public static object GetParameterDefaultValue(ParameterInfo parameterInfo, bool considerDefaultValueAttribute)
+        {
+            // In a scenario where the method signature is like "void Foo(Guid id = default(Guid))", 
+            // HasDefaultValue returns true, but the DefaultValue is null. This happens only in case of Reflection,
+            // whereas during execution time, calling the method "Foo()" does indeed supply the default value of
+            // 00000000-0000-0000-0000-000000000000 to the parameter "id".
+            if (parameterInfo.HasDefaultValue && parameterInfo.DefaultValue != null)
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ParameterDefaultValues.cs]: @pakrym found some shennanigans around default values when he was doing DI stuff. See https://github.com/aspnet/DependencyInjection/blob/dev/shared/Microsoft.Extensions.ParameterDefaultValue.Sources/ParameterDefaultValue.cs. We might need some of those here.

Comment (pakrym) [reply src/Microsoft.AspNetCore.Mvc.Core/Internal/ParameterDefaultValues.cs]: Yep, please use that thing.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ParameterDefaultValues.cs [hunk 2]
```diff
@@ -21,34 +22,26 @@ public static object[] GetParameterDefaultValues(MethodInfo methodInfo)
...
-                    }
-                    else
-                    {
-                        defaultValue = defaultValueAttribute.Value;
-                    }
-                }
+        public static object GetParameterDefaultValue(ParameterInfo parameterInfo)
+        {
+            object defaultValue;
+            if (!ParameterDefaultValue.TryGetDefaultValue(parameterInfo, out defaultValue))
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ParameterDefaultValues.cs]: `out var defaultValue` for 💯 points

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ParameterDefaultValues.cs [hunk 3]
```diff
@@ -21,34 +22,26 @@ public static object[] GetParameterDefaultValues(MethodInfo methodInfo)
...
+            {
+                var defaultValueAttribute = parameterInfo.GetCustomAttribute<DefaultValueAttribute>(inherit: false);
+                defaultValue = defaultValueAttribute?.Value;
 
-                values[i] = defaultValue;
+                if (defaultValue == null && parameterInfo.ParameterType.GetTypeInfo().IsValueType)
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ParameterDefaultValues.cs]: You could drop the `GetTypeInfo()`

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/ParameterDefaultValuesTest.cs [hunk 4]
```diff
@@ -25,6 +26,21 @@ public void GetParameterDefaultValues_ReturnsExpectedValues(string methodName, o
             Assert.Equal(expectedValues, actualValues);
         }
 
+        [Fact]
+        public void GetParameterDefaultValues_ReturnsExpectedValues_ForStructTypes()
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/ParameterDefaultValuesTest.cs]: `ForValueTypes`

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs [hunk 5]
```diff
@@ -194,13 +195,9 @@ private Task BindArgumentsAsync()
...
                 }
-                else if (parameter.ParameterInfo.HasDefaultValue)
+                else
                 {
-                    value = parameter.ParameterInfo.DefaultValue;
-                }
-                else if (parameter.ParameterInfo.ParameterType.IsValueType)
-                {
-                    value = Activator.CreateInstance(parameter.ParameterInfo.ParameterType);
+                    ParameterDefaultValue.TryGetDefaultValue(parameter.ParameterInfo, out value);
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs]: Does this need to account for value types when `TryGetValue` returns `false`?

Comment (kichalla) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs]: Good catch. Fixed it and added tests now.

**Pull request closed without merging**